src/npyiter.rs

```diff
@@ -0,0 +1,151 @@
+            DontNegateStrides => NPY_ITER_DONT_NEGATE_STRIDES,
+            CopyIfOverlap => NPY_ITER_COPY_IF_OVERLAP,
+            ReadWrite => NPY_ITER_READWRITE,
+            ReadOnly => NPY_ITER_READONLY,
+            WriteOnly => NPY_ITER_WRITEONLY,
+        }
+    }
+}
+
+pub struct NpyIterBuilder<'py, T> {
+    flags: npy_uint32,
+    array: &'py PyArrayDyn<T>,
+}
+
+impl<'py, T: TypeNum> NpyIterBuilder<'py, T> {
+    pub fn new<D: ndarray::Dimension>(array: &'py PyArray<T, D>) -> NpyIterBuilder<'py, T> {
+        NpyIterBuilder {
+            flags: 0,
+            array: array.into_dyn(),
+        }
+    }
+
+    pub fn add(mut self, flag: NpyIterFlag) -> Self {
+        self.flags |= flag.to_c_enum();
+        self
+    }
+
+    pub fn remove(mut self, flag: NpyIterFlag) -> Self {
+        self.flags &= !flag.to_c_enum();
+        self
+    }
+
+    pub fn build(self) -> PyResult<NpyIterSingleArray<'py, T>> {
+        let iter_ptr = unsafe {
+            PY_ARRAY_API.NpyIter_New(
+                self.array.as_array_ptr(),
+                self.flags,
+                NPY_ORDER::NPY_ANYORDER,
+                NPY_CASTING::NPY_SAFE_CASTING,
+                ptr::null_mut(),
+            )
+        };
+        let py = self.array.py();
+        NpyIterSingleArray::new(iter_ptr, py).ok_or_else(|| PyErr::fetch(py))
+    }
+}
+
+pub struct NpyIterSingleArray<'py, T> {
+    iterator: ptr::NonNull<objects::NpyIter>,
+    iternext: unsafe extern "C" fn(*mut objects::NpyIter) -> c_int,
+    empty: bool,
+    dataptr: *mut *mut c_char,
+
+    return_type: PhantomData<T>,
+    _py: Python<'py>,
+}
+
+impl<'py, T> NpyIterSingleArray<'py, T> {
+    fn new(iterator: *mut objects::NpyIter, py: Python<'py>) -> Option<NpyIterSingleArray<'py, T>> {
+        let mut iterator = ptr::NonNull::new(iterator)?;
+
+        // TODO replace the null second arg with something correct.
+        let iternext =
+            unsafe { PY_ARRAY_API.NpyIter_GetIterNext(iterator.as_mut(), ptr::null_mut())? };
+        let dataptr = unsafe { PY_ARRAY_API.NpyIter_GetDataPtrArray(iterator.as_mut()) };
+
+        if dataptr.is_null() {
+            unsafe { PY_ARRAY_API.NpyIter_Deallocate(iterator.as_mut()) };
+        }
+
+        Some(NpyIterSingleArray {
+            iterator,
+            iternext,
+            empty: false, // TODO: Handle empty iterators
+            dataptr,
+            return_type: PhantomData,
+            _py: py,
+        })
+    }
+}
+
+impl<'py, T> Drop for NpyIterSingleArray<'py, T> {
+    fn drop(&mut self) {
+        let _success = unsafe { PY_ARRAY_API.NpyIter_Deallocate(self.iterator.as_mut()) };
+        // TODO: Handle _success somehow?
+    }
+}
+
+impl<'py, T: 'py> std::iter::Iterator for NpyIterSingleArray<'py, T> {
+    type Item = &'py T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.empty {
+            None
+        } else {
+            let retval = Some(unsafe { &*(*self.dataptr as *mut T) });
```

Comment:
Now it looks like that this iterator returns the same value per iteration.
Maybe we have to move the pointer using stride?

Reply:
Actually, we don't need to ever move the pointer for the SingleArray case if the ExternalLoop flag is not set, best I can tell. The "iternext" function updates the data pointed to. Hence the double pointer. I'll write some more tests and confirm this though.